### PR TITLE
Single Result Objects for Public Methods

### PR DESCRIPTION
**Pull request by sarahkoop**
### Summary of changes

 - Update remaining public methods (American Express, Data Collector, Visa Checkout, Google Pay) to return single result objects to callbaks

 ### Checklist

 - [x] Added a changelog entry

### Authors

- @sarahkoop

### CODE REVIEW

**Review by tdchow — APPROVED** [review]
Looks good! just had a nullability question for one of the fields.

File: SEPADirectDebit/src/main/java/com/braintreepayments/api/SEPADirectDebitPaymentAuthRequestCallback.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ * [SEPADirectDebitClient.createPaymentAuthRequest].
+ */
+fun interface SEPADirectDebitPaymentAuthRequestCallback {
+    /**
+     * @param paymentAuthRequest the result of the SEPA create mandate call. If a nonce is present,
+     * no web-based mandate is required. If a nonce is not present, you
+     * must trigger the web-based mandate flow via
+     * [SEPADirectDebitLauncher.launch]
+     */
+    fun onResult(paymentAuthRequest: SEPADirectDebitPaymentAuthRequest)
```
Comment (saperi22) [inline SEPADirectDebit/src/main/java/com/braintreepayments/api/SEPADirectDebitPaymentAuthRequestCallback.kt:14]: do we send a request on a `onResult` callback? 🤔

File: AmericanExpress/src/test/java/com/braintreepayments/api/AmericanExpressClientUnitTest.java [hunk 2]
```diff
@@ -114,6 +122,27 @@ public void getRewardsBalance_callsListenerWithRewardsBalanceWithErrorCode_OnIns
         assertEquals("Insufficient points on card", rewardsBalance.getErrorMessage());
     }
 
+    @Test
+    public void getRewardsBalance_callsBackFailure_OnHttpError() {
```
Comment (hollabaq86) [inline AmericanExpress/src/test/java/com/braintreepayments/api/AmericanExpressClientUnitTest.java:126]: love that we're adding this test coverage. Question: do we have testing around analytic events being sent?

File: DataCollector/src/main/java/com/braintreepayments/api/DataCollector.java [hunk 3]
```diff
@@ -112,30 +112,27 @@ public void collectDeviceData(@NonNull final Context context, @NonNull final Dat
...
-                            request.setRiskCorrelationId(riskCorrelationId);
-                        }
-
-                        String correlationId =
-                                magnesInternalClient.getClientMetadataId(context, configuration, request);
-                        if (!TextUtils.isEmpty(correlationId)) {
-                            deviceData.put(CORRELATION_ID_KEY, correlationId);
-                        }
-                    } catch (JSONException ignored) {
+        braintreeClient.getConfiguration((configuration, error) -> {
```
Comment (hollabaq86) [inline DataCollector/src/main/java/com/braintreepayments/api/DataCollector.java:115]: does this work need to be in the CHANGELOG?

File: AmericanExpress/src/main/java/com/braintreepayments/api/AmericanExpressGetRewardsBalanceCallback.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * Callback for receiving result of
+ * [AmericanExpressClient.getRewardsBalance].
+ */
+fun interface AmericanExpressGetRewardsBalanceCallback {
+    /**
+     * @param americanExpressResult the [AmericanExpressResult]
+     */
+    fun onAmericanExpressResult(americanExpressResult: AmericanExpressResult?)
```
[the remark this answers is not in the document]
Comment (sarahkoop) [reply AmericanExpress/src/main/java/com/braintreepayments/api/AmericanExpressGetRewardsBalanceCallback.kt]: Nope that was just an oversight - they should all be non-null results good catch!

**Review by hollabaq86 — CHANGES_REQUESTED** [review]
This is a lot for one PR. There's 21 commits and 46 files changed...

Can we split this up into smaller PRs based on the payment method getting updated?

I know it may seem like more work up front, but it will be easier and faster for me and others to review a PR for each respected namespace (i.e. data collector, visacheckout, amex, etc).

Plus, if we need to revert a change during your major version work, the code is broken into smaller PRs.

File: SEPADirectDebit/src/main/java/com/braintreepayments/api/SEPADirectDebitPaymentAuthRequestCallback.kt (hunk 1 above)
Comment (sarahkoop) [reply SEPADirectDebit/src/main/java/com/braintreepayments/api/SEPADirectDebitPaymentAuthRequestCallback.kt]: That's actually a good call - we want to update all these callbacks from `onResult` to `onWhateverItReturns` - fixed this one

**Comment by sarahkoop** [comment]
Closing this replaced by #858 #859 #861 #862

**Pull request closed without merging**